WalletWasabi/BitcoinCore/Monitoring/P2pReconnector.cs

```diff
@@ -12,47 +12,41 @@ public class P2pReconnector : PeriodicRunner
 		public P2pReconnector(TimeSpan period, P2pNode p2pNode) : base(period)
 		{
 			P2pNode = Guard.NotNull(nameof(p2pNode), p2pNode);
-			Success = new TaskCompletionSource<object>();
+			Success = new TaskCompletionSource<bool>();
 		}
 
 		public P2pNode P2pNode { get; set; }
-		private TaskCompletionSource<object> Success { get; }
+		private TaskCompletionSource<bool> Success { get; }
 
 		protected override async Task ActionAsync(CancellationToken cancel)
 		{
-			try
+			Logger.LogInfo("Trying to reconnect to P2P...");
+			if (await P2pNode.TryDisconnectAsync(cancel).ConfigureAwait(false))
 			{
-				Logger.LogInfo("Trying to reconnect to P2P...");
-				P2pNode.Disconnect();
 				await P2pNode.ConnectAsync(cancel).ConfigureAwait(false);
-			}
-			catch
-			{
-				P2pNode.Disconnect();
-				throw;
-			}
 
-			Logger.LogInfo("Successfully reconnected to P2P.");
-
-			Success.TrySetResult(null);
+				Logger.LogInfo("Successfully reconnected to P2P.");
+				Success.TrySetResult(true);
+			}
 		}
 
 		public async Task StartAndAwaitReconnectionAsync(CancellationToken cancel)
 		{
 			await StartAsync(cancel).ConfigureAwait(false);
-			using var ctr = cancel.Register(() => Success.SetResult(null));
+			using var ctr = cancel.Register(() => Success.SetResult(false));
 			await Success.Task.ConfigureAwait(false);
 
 			try
 			{
-				using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(21));
+				using var cts = new CancellationTokenSource(Period * 2);
```

Comment:
Something like this, probably:

```csharp
using var cts = new CancellationTokenSource(Period * 2);
using var linked = CancellationTokenSource.CreateLinkedTokenSource(cancel, cts.Token);

// Stop the PeriodicRunner.
await StopAsync(linked.Token).ConfigureAwait(false);
```

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/4707/files#r524382818